clutel/clutel.go

```diff
@@ -32,3 +34,88 @@ func AddToOTELHTTPLabeler(
 		node.AddToOTELHTTPLabeler(labeler),
 	))
 }
+
+// GetSpan retrieves the current OpenTelemetry span from the context.
+func GetSpan(ctx context.Context) trace.Span {
+	return trace.SpanFromContext(ctx)
+}
+
+type spanBuilder struct {
+	kvs  map[string]any
+	opts []trace.SpanStartOption
+}
+
+// NewSpan produces a span builder that allows complete configuration and
+// attribution of the span before it gets started.
+func NewSpan() *spanBuilder {
+	return &spanBuilder{}
+}
+
+// WithAttrs adds attrs to the span.
+func (sb *spanBuilder) WithAttrs(kvs ...any) *spanBuilder {
+	if sb.kvs == nil {
+		sb.kvs = make(map[string]any)
+	}
+
+	sb.kvs = stringify.Normalize(kvs...)
+
+	return sb
+}
+
+// WithOpts configures the span with specific otel options.  Note that one
+// of the options can be `trace.WithAttributes()`; don't use that, use
+// spanBuilder.WithAttrs() instead.
+func (sb *spanBuilder) WithOpts(opts ...trace.SpanStartOption) *spanBuilder {
+	if sb.opts == nil {
+		sb.opts = make([]trace.SpanStartOption, 0, len(opts))
+	}
+
+	sb.opts = append(sb.opts, opts...)
+
+	return sb
+}
+
+// Start begins the span with the provided name and attaches it to the context.
+func (sb *spanBuilder) Start(
+	ctx context.Context,
+	name string,
+) context.Context {
+	nc := node.FromCtx(ctx)
+	if nc == nil {
+		return ctx
+	}
+
+	ctx, spanned := nc.AddSpan(ctx, name, sb.opts...)
+
+	if len(sb.kvs) > 0 {
+		spanned.ID = name
+		spanned = spanned.AddValues(ctx, sb.kvs)
+	} else {
+		spanned = spanned.AppendToTree(name)
+	}
```

Comment:
looking at this, I'm actually a little confused as to why we need a branch here. It seems in either case, a new node is created to hold the span and then, depending on whether we have KVs or not, we either set the name of the new node and then create another new node for the KVs or we create a new node for the name. Is there a reason we can't just always set the name on the new node with the span and then only create a second new node if there's KVs to add?

```suggestion
	spanned.ID = name

	if len(sb.kvs) > 0 {
		spanned = spanned.AddValues(ctx, sb.kvs)
	}
```

Reply:
yeah, this is fine.  I'll fix it up.